File: common/src/main/java/generations/gg/generations/core/generationscore/common/client/render/block/entity/GeneralUseBlockEntityRenderer.java
package generations.gg.generations.core.generationscore.common.client.render.block.entity;

import com.mojang.blaze3d.vertex.PoseStack;
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockAnimatedObjectInstance;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockLightValueProvider;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.BlockObjectInstance;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.ModelRegistry;
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.animation.FixedFrameAnimationInstance;
import generations.gg.generations.core.generationscore.common.world.level.block.entities.ModelProvidingBlockEntity;
import generations.gg.generations.core.generationscore.common.world.level.block.generic.GenericModelBlock;
import gg.generations.rarecandy.renderer.animation.AnimationInstance;
import gg.generations.rarecandy.renderer.rendering.ObjectInstance;
import gg.generations.rarecandy.renderer.storage.AnimatedObjectInstance;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public class GeneralUseBlockEntityRenderer<T extends ModelProvidingBlockEntity> implements BlockEntityRenderer<T> {

    public GeneralUseBlockEntityRenderer(BlockEntityRendererProvider.Context ctx) {
    }

    @Override
    public void render(@NotNull T blockEntity, float partialTick, PoseStack stack, @NotNull MultiBufferSource bufferSource, int packedLight, int packedOverlay) {
        if (!(blockEntity.getBlockState().getBlock() instanceof GenericModelBlock<?> block && block.canRender(blockEntity.getLevel(), blockEntity.getBlockPos(), blockEntity.getBlockState()))) return;
        stack.pushPose();
        ModelRegistry.prepForBER(stack, blockEntity);
        renderModels(stack, bufferSource, blockEntity, packedLight);
        stack.popPose();
    }

    protected void renderModels(PoseStack stack, MultiBufferSource buffersource, T blockEntity, int packedLight) {
        if(blockEntity.isAnimated()) renderModelFrameProvider(stack, buffersource, blockEntity, packedLight);
        else renderModelProvider(stack, buffersource, blockEntity, packedLight);
    }

    protected void renderModelProvider(PoseStack stack, MultiBufferSource buffersource, ModelProvidingBlockEntity blockEntity, int packedLight) {
        var model = ModelRegistry.get(blockEntity);

        if(model == null || model.renderObject == null) return;

        stack.scale(model.renderObject.scale, model.renderObject.scale, model.renderObject.scale);

        if (blockEntity.objectInstance == null) {
            int amount = instanceAmount();
            blockEntity.objectInstance = new ObjectInstance[amount];

            for (int i = 0; i < amount; i++)
                blockEntity.objectInstance[i] = blockEntity.generateInstance();
        }

        String variant = blockEntity.getVariant();

        for (var instance : blockEntity.objectInstance) {
            if (!Objects.equals(instance.materialId(), variant)) {
                instance.setVariant(variant);
            }

            instance.viewMatrix().set(stack.last().pose());
            ((BlockObjectInstance) instance).setLight(packedLight);
            if(blockEntity instanceof ModelContextProviders.TintProvider provider) ((BlockObjectInstance) instance).setTint(provider.getTint());
            model.render(instance, buffersource);
        }
    }

    protected int instanceAmount() {
        return 1;
    }

    protected void renderModelFrameProvider(PoseStack stack, MultiBufferSource buffersource, ModelProvidingBlockEntity blockEntity, int packedLight) {
        //TODO: Get this operational
        var model = ModelRegistry.get(blockEntity);

        if(model == null || model.renderObject == null) return;

        stack.scale(model.renderObject.scale, model.renderObject.scale, model.renderObject.scale);

        var amount = instanceAmount();
        blockEntity.objectInstance = new ObjectInstance[amount];

        for (int i = 0; i < amount; i++) {
            blockEntity.objectInstance[i] = blockEntity.generateInstance();
        }

        var primeInstance = blockEntity.objectInstance[0];

        if (model.renderObject.isReady()) {
            primeInstance.link(model.renderObject);

            var animationInstance = ((AnimatedObjectInstance) primeInstance);
            var animation = animationInstance.getAnimationsIfAvailable().get(blockEntity.getAnimation());

            if (animation != null) {
                if (blockEntity instanceof ModelContextProviders.FrameProvider frameProvider) {
                    animationInstance.changeAnimation(new FixedFrameAnimationInstance(animation, frameProvider.getFrame()));
                } else {
                    animationInstance.changeAnimation(new AnimationInstance(animation));
                }
            }
        }

        primeInstance.viewMatrix().set(stack.last().pose());
        ((BlockLightValueProvider) primeInstance).setLight(packedLight);
        if(blockEntity instanceof ModelContextProviders.TintProvider provider) ((BlockAnimatedObjectInstance) primeInstance).setTint(provider.getTint());

        var instance = (AnimatedObjectInstance) primeInstance;

        if(blockEntity instanceof ModelContextProviders.FrameProvider frameProvider && instance.currentAnimation instanceof FixedFrameAnimationInstance fixedAnimation && fixedAnimation.getCurrentTime() != frameProvider.getFrame()) {
            fixedAnimation.setCurrentTime(frameProvider.getFrame());
        }

        model.render(instance, buffersource);
    }

    protected void renderResourceLocation(MultiBufferSource source, ResourceLocation location, PoseStack stack, ObjectInstance objectInstance) {
        objectInstance.transformationMatrix().set(stack.last().pose());

        var model = ModelRegistry.get(location);
        if(model != null) model.render(objectInstance, source);
    }

    @Override
    public boolean shouldRenderOffScreen(T blockEntity) {
        return true;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/client/render/block/entity/GeneralUseBlockEntityRenderer.kt
package generations.gg.generations.core.generationscore.common.client.render.block.entity

import com.cobblemon.mod.common.util.toVec3d
import com.mojang.blaze3d.vertex.PoseStack
import generations.gg.generations.core.generationscore.common.client.model.InstanceProvider
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.AngleProvider
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.FrameProvider
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.ModelProvider
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.TintProvider
import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.VariantProvider
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.*
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.ModelRegistry.prepForBER
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.animation.FixedFrameAnimationInstance
import generations.gg.generations.core.generationscore.common.world.level.block.entities.ModelProvidingBlockEntity
import generations.gg.generations.core.generationscore.common.world.level.block.generic.GenericModelBlock
import gg.generations.rarecandy.renderer.animation.AnimationInstance
import gg.generations.rarecandy.renderer.rendering.ObjectInstance
import gg.generations.rarecandy.renderer.storage.AnimatedObjectInstance
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.entity.BlockEntity

open class GeneralUseBlockEntityRenderer<T>(ctx: BlockEntityRendererProvider.Context) : BlockEntityRenderer<T> where T : BlockEntity, T: VariantProvider, T: ModelProvider, T: InstanceProvider {


    override fun render(
        blockEntity: T,
        partialTick: Float,
        stack: PoseStack,
        bufferSource: MultiBufferSource,
        packedLight: Int,
        packedOverlay: Int
    ) {

        blockEntity.blockState.block.instanceOrNull<GenericModelBlock>()?.takeIf { it.canRender(blockEntity) } ?: return

        if (blockEntity.instanceArray == null) {
            val amount = instanceAmount()
            blockEntity.instanceArray = arrayOfNulls(amount)

            for (i in 0 until amount) blockEntity.instanceArray!![i] = blockEntity.generateInstance()
        }

        stack.pushPose()
        if(blockEntity is AngleProvider) prepForBER(stack, blockEntity)
        renderModels(stack, bufferSource, blockEntity, packedLight)
        stack.popPose()
    }

    protected open fun renderModels(
        stack: PoseStack,
        buffersource: MultiBufferSource,
        blockEntity: T,
        packedLight: Int
    ) {
        if (blockEntity.isAnimated) renderModelFrameProvider(stack, buffersource, blockEntity, packedLight)
        else renderModelProvider(stack, buffersource, blockEntity, packedLight)
    }

    protected fun renderModelProvider(
        stack: PoseStack,
        buffersource: MultiBufferSource,
        blockEntity: T,
        packedLight: Int
    ) {
        val model = ModelRegistry[blockEntity]

        if (model?.renderObject == null) return

        stack.scale(model.renderObject!!.scale, model.renderObject!!.scale, model.renderObject!!.scale)

        val variant = blockEntity.variant

        blockEntity.instanceArray!!.requireNoNulls().forEach { instance ->
            if (instance.materialId() != variant) {
                instance.setVariant(variant)
            }

            instance.transformationMatrix().set(stack.last().pose())

            (instance as BlockObjectInstance).light = packedLight
            if (blockEntity is TintProvider) instance.tint = blockEntity.tint
            model.render(instance, buffersource)
        }
    }

    protected open fun instanceAmount(): Int {
        return 1
    }

    protected fun renderModelFrameProvider(
        stack: PoseStack,
        buffersource: MultiBufferSource,
        blockEntity: T,
        packedLight: Int
    ) {
        //TODO: Get this operational
        val model = ModelRegistry[blockEntity]

        if (model?.renderObject == null) return

        stack.scale(model.renderObject!!.scale, model.renderObject!!.scale, model.renderObject!!.scale)

        val primeInstance = blockEntity.instanceArray!![0]!!

        if (model.renderObject!!.isReady) {
            primeInstance.link(model.renderObject)

            val animationInstance = (primeInstance as AnimatedObjectInstance)
            val animation = animationInstance.animationsIfAvailable[blockEntity.animation]

            if (animation != null) {
                if (blockEntity is FrameProvider) {
                    animationInstance.changeAnimation(FixedFrameAnimationInstance(animation, blockEntity.frame))
                } else {
                    animationInstance.changeAnimation(AnimationInstance(animation))
                }
            }
        }

        val offset = blockEntity.blockPos.toVec3d().subtract(Minecraft.getInstance().cameraEntity!!.position())

        primeInstance.transformationMatrix().set(stack.last().pose()).translate(offset.x.toFloat(),
            offset.y.toFloat(), offset.z.toFloat()
        )
        (primeInstance as BlockLightValueProvider).light = packedLight
        if (blockEntity is TintProvider) (primeInstance as BlockAnimatedObjectInstance).tint = blockEntity.tint

        val instance = primeInstance as AnimatedObjectInstance

        val provider = blockEntity.instanceOrNull<FrameProvider>()

        if(provider != null) instance.currentAnimation?.instanceOrNull<FixedFrameAnimationInstance>()?.takeIf { it.currentTime != provider.frame }?.run { this.currentTime = provider.frame }

        model.render(instance, buffersource)
    }

    protected fun renderResourceLocation(
        source: MultiBufferSource,
        location: ResourceLocation,
        stack: PoseStack,
        objectInstance: ObjectInstance
    ) {
        objectInstance.transformationMatrix().set(stack.last().pose())

        val model = ModelRegistry[location]
        model?.render(objectInstance, source)
    }

    override fun shouldRenderOffScreen(blockEntity: T): Boolean {
        return true
    }
}
